### DISCUSSION TITLE
CustomAttributes cache per script engine

### DISCUSSION THREAD

**Discussion by ebrandin** [ClearFoundry/ClearScript#540] [Q&A] [answered]
I am trying to use a CustomAttributeLoader to implement API security level and version control, where every user/script may have a different set of available exposed types and methods. As of my understand, this is not quite possible since the cache in CustomAttributes is global and the first executed script will cache the custom attributes. Clearing the cache before every script would add a great performance penalty, and also would not quite work in a multi tenant application.

Currently the only way forward is to add a specific access control method to every exposed method. That would however not be possible if one would want to add security levels to .NET classes/methods.

If a ScriptEngine optionally could contain a local CustomAttributes cache, it would be possible to create pools of engines, differentiated by their capabilities.

Is there another way to solve this use case?

**Comment by ClearScriptLib** [comment]
Hi @ebrandin,

>I am trying to use a CustomAttributeLoader to implement API security level and version control, where every user/script may have a different set of available exposed types and methods.

The title mentions per-engine caching, but here you refer to users and scripts. Can you describe the relationship between users, scripts, and engines in your architecture?

>If a ScriptEngine optionally could contain a local CustomAttributes cache, it would be possible to create pools of engines, differentiated by their capabilities.

Per-engine attribute caching seems doable, but it's a departure from standard .NET attribute semantics, which support one static set of attributes per resource. That _might_ be OK, but we'd prefer to explore other options first.

>Is there another way to solve this use case?

It seems like you wish to expose your entire API to all clients, and then use attributes to rope off restricted areas. Would it not be possible instead to expose only the portions each client is allowed to access?

Thanks!

**Reply by ebrandin** [reply]
We are building a multi-tenant application where all users have access to the full API. However, in order to facilitate backward/forward compatibility and trust, every script needs to be tagged with an API version and a set of capabilities.

Instead of using the built in `ScriptUsageAttribute`, we want to use a `CustomAttributeLoader` in combination with our own custom attribute that contain version and capability information. Note then that the outcome of the custom attribute loader will depend on information passed the application and thus unsuitable to be cached globally.

The way we envision caching to work is that the cache is an object can be created explicitly when needed (in our case one cache per API version and capability) and then passed into a `ScriptEngine`. When ClearScript checks for script access in `CustomAttributes.GetOrLoad`, such an explicitly set cache would take precedence over the global/static cache.

I believe that this would not conflict with the .NET attribute semantics, nor would the performance penalty be significant as long as the number caches is kept to a reasonable amount. This would greatly extend the possible use cases of the custom attribute loader.

I hope that I was able to make our use case clear enough.
Thanks!

**Reply by ClearScriptLib** [reply]
>The way we envision caching to work is that the cache is an object can be created explicitly when needed (in our case one cache per API version and capability) and then passed into a `ScriptEngine`.

In other words:

- Each script engine optionally holds a reference to an attribute cache,
- An attribute cache can be shared among multiple script engines.
- A script engine does _not_ need to handle more than one attribute cache.

Is all that that correct?

>When ClearScript checks for script access in `CustomAttributes.GetOrLoad`, such an explicitly set cache would take precedence over the global/static cache.

But you aren't asking for ClearScript to support any new attributes, are you? That is, since you'll be using your own attributes, won't you need to check for them yourself? Or will your attributes be based on `ScriptUsageAttribute`?

**Reply by ebrandin** [reply]
> Is all that that correct?

Yes on all 3 bullet points.

> But you aren't asking for ClearScript to support any new attributes, are you?

No, we are note asking for any new attributes. We are not basing our own attributes on `ScriptUsageAttribute` either, but rather returning a list containing an appropriate `ScriptUsageAttribute` created by our `CustomAttributeLoader` depending on whether the entry is allowed or not.

**Reply by ClearScriptLib** [reply]
A few more questions:

1. Above, you mentioned custom (non-ClearScript) attributes that hold version and capability information. Can you clarify what these attributes would be applied to, and how they'd be used in your scenario?

2. In addition to `Clear`, what operations on the cache object would be useful?

3. `LoadCustomAttributes` is invoked in the context of an internal lock. According to the [documentation](https://clearscript.clearfoundry.net/Reference/html/M_Microsoft_ClearScript_CustomAttributeLoader_LoadCustomAttributes__1.htm): "This method is performance-critical. Overrides must not invoke script engine methods or other ClearScript functionality." Would that requirement be a problem for your override?

Thanks!

**Reply by ebrandin** [reply]
1. All API methods and types are in our application tagged with a version, and so is every script. The CustomAttributeLoader will ensure that only the API avaliable for the requested version will be available at runtime. This makes i possible to ensure proper version compatibility of a script.
Capabilies are used very much like any smart phone app asking for access to certain OS functionality, like camera and contacts. A portion of the API will via our attribute be tagged with such extendend capability. A script that requries such API will need to declare that along with version. The application will require the end user to accept this before running the script.
A script that fails to declare the correct version or capability should fail at runtime.

2. No methods other than constructor (or creation method) and a setter on the `ScriptEngine `is required as I see it. I don't even see that `Clear `is needed if we can attach a cache to a `ScriptEngine`. 

3. The only thing the custom attribute loader does is to find our application specific attributes and compares the version and an optional extended capability with the version and capability specified by the script. There are no external or time consuming operation performed.

**Reply by ClearScriptLib** [reply, marked as answer]
Thanks. We'll look into this for an upcoming release.

**Reply by ebrandin** [reply]
Thank you, it would be great if such a feature could make it into the product soon.

**Reply by ClearScriptLib** [reply]
Hi @ebrandin,

One more question: In your custom attribute loader, how do you intend to determine which engine (or tenant, or application, etc.) is making the request?

Thanks!

**Reply by ebrandin** [reply]
The plan is to use thread static variables on our custom attribute loader to transfer version and capability information. Would this work or are there situations where the custom attribute loader may be invoked on a different thread than thread invoking `ScriptEngine.Execute`?

**Reply by ClearScriptLib** [reply]
> Would this work or are there situations where the custom attribute loader may be invoked on a different thread than thread invoking `ScriptEngine.Execute`?

Methods like `Execute` are just one way to enter the script engine. Directly calling a script function is another way. Even retrieving a script object property could involve executing a "getter" in script code. As for threads, a script could [`await`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/await) a promise bound to a .NET task, and if that task completes on another thread, the script engine could be invoked there as well.

Your custom attribute loader could use [`ScriptEngine.Current`](https://clearscript.clearfoundry.net/Reference/html/P_Microsoft_ClearScript_ScriptEngine_Current.htm) and map the result to the required data. We could add something like `ScriptEngine.HostData` to facilitate that. Another possibility might be to allow per-engine custom attribute loaders. We're still working out the API details, and we welcome suggestions!

**Reply by ebrandin** [reply]
Thank you for looking into this.

The idea of a per engine custom attribute loader, rather than a per engine cache, seems like a very intereseting idea which may solve all issues nicely. If I understand this idea correctly this could mean:

1. The attribute cache and the needed logic would move from the static CustomAttributes to instance methods of CustomAttributeLoader, enabling multiple caches.
2. An implementation of CustomAttributeLoader can hold the state needed to determine access to an API, in our case verision and capability, thus there would be no need to use thread static variables like we planned.
3. The cache would be an internal state in a CustomAttributeLoader, and there would be no reason to expose this to the application.
4. A CustomAttributeLoader would need a new ClearCache method. The current `CustomAttributes.ClearCache` method could probably be implemented by calling `HostSettings.CustomAttributeLoader.ClearCache()`. Note that if all engines are created with a custom attribute loader, this would be a no-op since this static attribute loader would never have been filled. Any engine created without a custom attribute loader would use the old static custom attribute loader.

To me it seems that this could solve new use cases without breaking the existing API.

**Reply by ClearScriptLib** [reply]
Hi again,

Yes. Allowing per-engine loaders, each with a private cache, is probably the right thing. Exposing loaders and caches in a decoupled fashion could easily lead to unexpected behavior – e.g., the host could set up multiple loaders that share a cache.

>An implementation of CustomAttributeLoader can hold the state needed to determine access to an API, in our case verision and capability, thus there would be no need to use thread static variables like we planned.

Exactly. However, you originally envisioned multiple engines sharing a loader but using separate caches. Now we're talking about a separate loader with a private cache for each engine – or perhaps one for each level of API access your system supports. Will that still work for you?

>A CustomAttributeLoader would need a new ClearCache method.

We'd like to avoid that if possible. Currently there's a global cache that needs to be cleared when the host assigns a new global loader. With each loader having a private cache, that's no longer necessary.

Thoughts?

**Reply by ebrandin** [reply]
Hi,

> Now we're talking about a separate loader with a private cache for each engine

What I meant was that a loader could be specified (via contstructor or similar) when creating an engine, rather than via a static global loader like today. This would mean that the using application could share loaders between engines. In our case we would keep a dictionary of loaders per API access. This naturally also requires the cache to be thread safe, just like today.

> Will that still work for you?

A strict 1-1 mapping between engine and loader would probably not work well for us since we want to dispose an engine after every executed script. That would essentially nullify the cache altogether for us, probably with sever performance implications.

Finally, I agree that a `ClearCache` method is no longer needed. I never saw the use case to make this method public in the first place. I just wanted to point out that a new solution could be made backward compatible for other users that might have found other use cases for that method.

**Reply by ClearScriptLib** [reply]
OK, great! We seem to be in agreement on the requirements and API design. Our next release should be out soon after V8 11.8 goes stable, so 2-3 weeks from now. Thanks for helping us enhance ClearScript!

**Reply by ClearScriptLib** [reply]
[Version 7.4.4](https://github.com/microsoft/ClearScript/releases/tag/7.4.4) added [`ScriptEngine.CustomAttributeLoader`](https://clearscript.clearfoundry.net/Reference/html/P_Microsoft_ClearScript_ScriptEngine_CustomAttributeLoader.htm).
